java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java

```diff
@@ -107,42 +117,89 @@ public String toString() {
     public Node parent;
     private final List<LearnedConstraint> learnedConstraints;
 
+    private final List<LearnedValue> learnedSymbols;
+
     Node(ProgramPoint programPoint, @Nullable ProgramState programState) {
       this.programPoint = programPoint;
       this.programState = programState;
       learnedConstraints = new ArrayList<>();
+      learnedSymbols = new ArrayList<>();
     }
 
     public void setParent(@Nullable Node parent) {
       this.parent = parent;
-      if(parent != null) {
+      if (parent != null) {
         programState.constraints.forEach((sv, c) -> {
-          if(parent.programState.getConstraint(sv) != c) {
-            learnedConstraints.add(new LearnedConstraint(sv, c));
+          if (parent.programState.getConstraint(sv) != c) {
+            addConstraint(sv, c);
+          }
+        });
+        programState.values.forEach((s, sv) -> {
+          if (parent.programState.getValue(s) != sv) {
+            learnedSymbols.add(new LearnedValue(sv, s));
           }
         });
       }
     }
 
-    public List<LearnedConstraint> learnedConstraints() {
+    private void addConstraint(SymbolicValue sv, Constraint constraint) {
+      // FIXME : this might end up adding twice the same SV in learned constraints. Safe because of find first in SECheck.flows
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1134)